design/catchup-and-audit-proposal.md

```diff
@@ -0,0 +1,156 @@
+# Catchup and audit design proposal
+
+## Why
+
+### Requirement 1: No synchronization issues between any ledgers
+
+**Problem:** When multiple ledgers have interdependencies current independent catch-up
+of different ledgers can break things as follows:
+- suppose we have ledgers A and B, and validation of transactions from A is dependent
+  on contents of B
+- suppose some node X started catch-up and caught up ledger A
+- after that pool ordered transaction A1, then B1
+- B1 would prohibit processing of A1 if B1 was ordered first
+- note that node X doesn't contain A1 at this point, since catch up of ledger A was
+  done before transaction A1 was ordered by pool
+- node X continues catch-up and catches up ledger B already containing B1
+- node X finishes catch up and tries to process stashed 3PC messages
+- it learns that it should process A1, however it already has B1 which cause
+  A1 to be rejected by node X
+- so node X falls out of consensus
+
+So, ledgers should be caught up consistently, that is, following previous example,
+if A1 wasn't caught up then B1 shouldn't be caught up as well.
+
+### Requirement 2: Auditability/history
+
+It should be possible to audit if any given transaction was written to the ledger
+legitimately, that is all validations really passed.
+
+Examples:
+- Validate authorization rules
+- Domain-config: multi-signature values
+- Pool-config: multi-signature values
+- Pool-domain: whether Steward(s) have rights to add/edit NODE txn
+
+The issue is caused by a fact, that  there is no way to restore order of transactions
+between ledgers after they are written, and hence no way to know what was the total state
+of the system at the moment of applying any given transaction to be able to audit if it
+was applied correctly.
+
+**Nice to have:** It should be _easily_ possible to audit if any given transaction was
+written to the ledger legitimately. In other words, it should be easily possible to get
+the current state of the system at the moment before the txn was applied.
+
+Input:
+- Transaction
+- Ledger(s)
+
+Output:
+- Results of post-write audit (off-ledger checks for validation against the state
+  at the moment when txn was written)
+
+### Requirement 3 (optional): Deterministically and consistently restore 3PC state (last ordered batch viewNo/ppSeqNo)
+
+Currently when enough nodes (but not all) restart simultaneously nodes that didn’t
+restart fall out of consensus. Some sequence of restarts may lead to full loss of
+consensus in pool. We have some partial fixes for this problem, but they are not
+reliable and edge cases still exist.
+
+## How: Add common ledger with links to other ledgers
+
+### Description
+
+Create one more ledger (let’s call it Audit), and for each ordered batch make
```

Comment:
Why create a separate ledger? Why not persist viewNo and ppSeqNo in the respective ledger apart from storing other ledger's root hash if it had changed. So for example domain ledger looks like this
```
Domain ledger Batch1
     (viewNo1, ppSeqNo1)
     pool root (part of domain ledger merkle tree as a leaf, like any other txn)
     config root (part of domain ledger merkle tree a leaf, like any other txn)
     domain txn 1
     domain txn 2
Domain ledger Batch2
     (viewNo1, ppSeqNo2)
     // no change in pool or config ledger
     domain txn 3
     domain txn 4
     domain txn 5
Domain ledger Batch4    // batch 3 was for config ledger
     (viewNo1, ppSeqNo4)
     // config ledger changed
     config ledger new root (part of domain ledger merkle tree a leaf, like any other txn)
     domain txn 6
     domain txn 7
```